domino_data/transfer.py

```diff
@@ -0,0 +1,82 @@
+from typing import BinaryIO, Generator, Optional
+
+import io
+import shutil
+import threading
+from concurrent.futures import ThreadPoolExecutor
+
+import urllib3
+
+MAX_WORKERS = 10
+MB = 2**20
+
+
+def split_range(start: int, end: int, step: int) -> Generator[tuple[int, int], None, None]:
+    """Yield all steps from start to end as tuple of integers.
+
+    All returned block are inclusive to ensure every integer is covered from start to end.
+
+    Example:
+        start: 0, end: 10, step: 3 -> (0, 2), (3, 6), (7, 10)
+        start: 0, end: 10, step: 4 -> (0, 3), (4, 8), (9, 10)
+
+    Args:
+        start: start of the range to split
+        end: end of the range to split
+        step: step size
+
+    Yields:
+        tuple[int, int]
+    """
+    r = range(start, end, step)
+    max_block = start
+
+    for min_block, max_block in zip(r[:-1], r[1:]):
+        yield (min_block, max_block - 1)
+
+    yield (max_block, end)
+
+
+class BlobTransfer:
+    def __init__(
+        self,
+        url: str,
+        destination: BinaryIO,
+        max_workers: int = MAX_WORKERS,
+        chunk_size: int = 16 * MB,
```

Comment:
where does the 16 come from?

Reply:
https://docs.aws.amazon.com/AmazonS3/latest/userguide/optimizing-performance-guidelines.html#optimizing-performance-guidelines-get-range